File: src/main/java/fisheye270/calibration/FisheyeModel.java
package fisheye270.calibration;

import org.opencv.calib3d.Calib3d;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

/**
 * OpenCV fisheye LUTs (perspective debug path) and Kannala–Brandt projection
 * that supports FOV &gt; 180° (z can be negative).
 *
 * Runtime stitching must NOT use the perspective maps — use cylindrical maps.
 */
public final class FisheyeModel {
    private FisheyeModel() {}

    public static Mat[] undistortMaps(Intrinsics intr, double balance, double fovScale) {
        Mat K = intr.K();
        Mat D = intr.D();
        Mat R = Mat.eye(3, 3, CvType.CV_64FC1);
        Mat P = new Mat();
        Size src = new Size(intr.width, intr.height);
        Calib3d.fisheye_estimateNewCameraMatrixForUndistortRectify(
                K, D, src, R, P, balance, src, fovScale);
        Mat map1 = new Mat();
        Mat map2 = new Mat();
        Calib3d.fisheye_initUndistortRectifyMap(
                K, D, R, P, src, CvType.CV_32FC1, map1, map2);
        K.release();
        D.release();
        R.release();
        P.release();
        return new Mat[]{map1, map2};
    }

    public static Mat remap(Mat src, Mat mapX, Mat mapY) {
        Mat dst = new Mat();
        Imgproc.remap(src, dst, mapX, mapY, Imgproc.INTER_LINEAR);
        return dst;
    }

    /**
     * Project a camera-frame ray to a fisheye pixel.
     * @return false if outside the image or beyond the assumed FOV
     */
    public static boolean project(Intrinsics intr, double x, double y, double z, float[] uv) {
        double r = Math.hypot(x, y);
        double theta = Math.atan2(r, z);
        double t2 = theta * theta;
        double t4 = t2 * t2;
        double scale = 1.0 + intr.k1 * t2 + intr.k2 * t4 + intr.k3 * t4 * t2 + intr.k4 * t4 * t4;
        double thetaD = theta * scale;
        double u;
        double v;
        if (r <= 1e-12) {
            u = intr.cx;
            v = intr.cy;
        } else {
            u = intr.fx * thetaD * x / r + intr.cx;
            v = intr.fy * thetaD * y / r + intr.cy;
        }
        double maxTheta = Math.toRadians(Math.min(Math.max(intr.fovHorizontalDeg * 0.55, 80.0), 170.0));
        if (!Double.isFinite(u) || !Double.isFinite(v) || theta >= maxTheta
                || u < 0 || v < 0 || u >= intr.width || v >= intr.height) {
            return false;
        }
        uv[0] = (float) u;
        uv[1] = (float) v;
        return true;
    }
}

File: fisheye_270/calibration/fisheye_model.py
"""OpenCV fisheye (Kannala-Brandt) projection and undistort LUTs.

cv2.fisheye.calibrate(object_points, image_points, image_size, K, D)
    -> RMS, K, D, rvecs, tvecs
    Requires a calibration target. Mode B only.

cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(K, D, size, R, P, balance)
    Chooses a rectilinear K' after undistortion.
    balance=0 crops to valid pixels; 1 keeps FOV (black borders).

cv2.fisheye.initUndistortRectifyMap(K, D, R, P, size, m1type)
    Builds mapX/mapY: dest rectilinear pixel -> source fisheye pixel.

cv2.remap(src, mapX, mapY, INTER_LINEAR)
    Runtime: one bilinear sample per output pixel. Maps are precomputed.

Why LUTs: a 1920x540 remap is a memory gather. Recomputing atan/polynomial
per frame is wasted work while K, D, and pose are static.
"""

from __future__ import annotations

from typing import Tuple

import cv2
import numpy as np

from .camera_model import Intrinsics


def undistort_maps(
    intr: Intrinsics,
    dst_size: Tuple[int, int] | None = None,
    balance: float = 0.5,
    fov_scale: float = 1.0,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Perspective (rectilinear) undistort maps. Debug / Phase 2 view only.

    A 190° fisheye cannot be shown as a single perspective image without
    either cropping FOV or stretching the rim to infinity. Do not stitch
    these frames for the 270° panorama — use cylindrical/spherical maps.
    """
    w, h = (dst_size if dst_size else (intr.width, intr.height))
    K, D = intr.as_opencv()
    R = np.eye(3, dtype=np.float64)
    new_K = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(
        K, D, (intr.width, intr.height), R, balance=float(balance),
        new_size=(w, h), fov_scale=float(fov_scale),
    )
    map1, map2 = cv2.fisheye.initUndistortRectifyMap(
        K, D, R, new_K, (w, h), cv2.CV_32FC1,
    )
    return map1, map2, new_K


def project_fisheye(rays_cam: np.ndarray, intr: Intrinsics) -> Tuple[np.ndarray, np.ndarray]:
    """Project camera-frame rays to fisheye pixels. Supports FOV > 180° (z < 0).

    theta = atan2(sqrt(x^2+y^2), z)
    theta_d = theta * (1 + k1 theta^2 + k2 theta^4 + k3 theta^6 + k4 theta^8)
    u = fx * (theta_d * x / r) + cx
    """
    x = rays_cam[..., 0]
    y = rays_cam[..., 1]
    z = rays_cam[..., 2]
    r = np.hypot(x, y)
    theta = np.arctan2(r, z)
    t2 = theta * theta
    t4 = t2 * t2
    t6 = t4 * t2
    t8 = t4 * t4
    scale = 1.0 + intr.k1 * t2 + intr.k2 * t4 + intr.k3 * t6 + intr.k4 * t8
    theta_d = theta * scale
    # rays behind the camera at exactly 180° have r~0 and theta=pi — invalid
    valid = (r > 1e-12) | (np.abs(theta) < 1e-8)
    inv_r = np.where(r > 1e-12, 1.0 / np.maximum(r, 1e-12), 0.0)
    u = intr.fx * theta_d * x * inv_r + intr.cx
    v = intr.fy * theta_d * y * inv_r + intr.cy
    u = np.where(r <= 1e-12, intr.cx, u)
    v = np.where(r <= 1e-12, intr.cy, v)
    in_image = (
        valid
        & np.isfinite(u)
        & np.isfinite(v)
        & (u >= 0)
        & (v >= 0)
        & (u < intr.width)
        & (v < intr.height)
        & (theta < np.deg2rad(min(max((intr.fov_horizontal_deg or 190.0) * 0.55, 80.0), 170.0)))
    )
    return np.stack([u, v], axis=-1).astype(np.float32), in_image


def remap(frame: np.ndarray, mapx: np.ndarray, mapy: np.ndarray) -> np.ndarray:
    return cv2.remap(
        frame, mapx, mapy, interpolation=cv2.INTER_LINEAR,
        borderMode=cv2.BORDER_CONSTANT, borderValue=(0, 0, 0),
    )
